File: ios/Sources/PairingCode.swift
/// A pairing code: the server to dial and the host keys it offers, read from a
/// scanned or tapped link. The rules and their order are `shared/pairing.ts`'s,
/// and `shared/pairing.swift.test.ts` holds this file to the same vectors
/// (remote.md §4b). It imports nothing, so that test can compile it on a Mac.
struct PairingCode: Equatable {
    let user: String
    /// A host name or an IPv4 address.
    let host: String
    /// 0 for sshd's default port, the way `ServerRecord.port` stores it.
    let port: Int
    /// `SHA256:…` as `ssh-keygen -lf` prints it, one per host key.
    let fingerprints: [String]

    var destination: String { "\(user)@\(host)" }

    enum Read: Equatable {
        case code(PairingCode)
        case problem(String)
    }

    static let version = 1
    static let maxFingerprints = 4

    enum Problem {
        static let notACode = "This is not a Ledge pairing code."
        static let newer = "This pairing code needs a newer version of Ledge. Update the app, then try the code again."
        static let version = "The pairing code's version is not one Ledge recognizes."
        static let encoding = "The pairing code has a damaged character in it."
        static let user = "The pairing code has no valid account name."
        static let host = "The pairing code has no valid host name or IP address."
        static let port = "The pairing code's port is not a whole number from 1 to 65535."
        static let noFingerprint = "The pairing code has no host key fingerprint."
        static let fingerprint = "The pairing code has a host key fingerprint that is not a SHA256 fingerprint."
        static let tooManyFingerprints = "The pairing code names more than \(maxFingerprints) host keys."
        static func repeated(_ field: String) -> String { "The pairing code gives the \(field) more than once." }
    }

    /// Everything here walks Unicode scalars, not `Character`s. A combining mark
    /// joins the character before it, so "&" followed by one is a `Character`
    /// that is not "&", and the TypeScript reader would still split there.
    static func read(_ text: String) -> Read {
        let scalars = trimmed(Array(text.unicodeScalars))
        guard let hash = scalars.firstIndex(of: "#") else { return .problem(Problem.notACode) }
        let base = asciiLowercased(string(scalars[..<hash]))
        guard base == "https://ledge.sh/pair" || base == "ledge://pair" else { return .problem(Problem.notACode) }

        let fields: [(key: String, raw: String)] = scalars[(hash + 1)...].split(separator: "&").map { segment in
            guard let eq = segment.firstIndex(of: "=") else { return (string(segment), "") }
            return (string(segment[..<eq]), string(segment[(eq + 1)...]))
        }
        func raws(_ key: String) -> [String] { fields.filter { $0.key == key }.map(\.raw) }

        let versions = raws("v")
        if versions.isEmpty { return .problem(Problem.notACode) }
        if versions.count > 1 { return .problem(Problem.repeated("version")) }
        guard let version = versionNumber(versions[0]) else { return .problem(Problem.version) }
        if version > Self.version { return .problem(Problem.newer) }

        for (key, name) in [("u", "account name"), ("h", "host"), ("p", "port")] where raws(key).count > 1 {
            return .problem(Problem.repeated(name))
        }
        var decoded: [String: [String]] = [:]
        for key in ["u", "h", "p", "k"] {
            let values = raws(key).map(percentDecoded)
            if values.contains(where: { $0 == nil }) { return .problem(Problem.encoding) }
            decoded[key] = values.compactMap { $0 }
        }

        var fingerprints: [String] = []
        for fingerprint in decoded["k"] ?? [] where !fingerprints.contains(fingerprint) {
            fingerprints.append(fingerprint)
        }
        let code = PairingCode(
            user: decoded["u"]?.first ?? "",
            host: decoded["h"]?.first ?? "",
            port: portField(decoded["p"]?.first),
            fingerprints: fingerprints
        )
        if let problem = code.problem { return .problem(problem) }
        return .code(code)
    }

    /// A stored server, reduced to what the pin rule reads.
    struct Known: Equatable {
        let id: String
        let destination: String
        let port: Int
        /// The pinned key's `SHA256:…`, or "" for a record whose pin was dropped.
        let fingerprint: String
    }

    /// What the stored servers make of this code (remote.md §4b, the first rule).
    enum Match: Equatable {
        /// No record for this account at this host and port. The key the server
        /// offers has to be one the code names, and it is pinned.
        case new
        /// This account's record, pinned to a key the code names. The pin
        /// decides the dial, and nothing is pinned again.
        case pinned(id: String)
        /// This account's record with its pin dropped. The code's key is pinned.
        case unpinned(id: String)
        /// A record at this host and port is pinned to a key the code does not
        /// name. The code is refused, the way a changed host key is.
        case conflict(pinned: String)
    }

    /// A host key belongs to a host and a port rather than to an account, so a
    /// pin on any account there can refuse the code. A host name compares
    /// without regard to ASCII case, since DNS ignores it.
    func match(_ known: [Known]) -> Match {
        let host = Self.asciiLowercased(self.host)
        let here = known.filter { Self.asciiLowercased(Self.host(of: $0.destination)) == host && $0.port == port }
        if let stale = here.first(where: { !$0.fingerprint.isEmpty && !fingerprints.contains($0.fingerprint) }) {
            return .conflict(pinned: stale.fingerprint)
        }
        guard let mine = here.first(where: { Self.user(of: $0.destination) == user }) else { return .new }
        return mine.fingerprint.isEmpty ? .unpinned(id: mine.id) : .pinned(id: mine.id)
    }

    private static func user(of destination: String) -> String {
        String(destination.unicodeScalars.prefix(while: { $0 != "@" }))
    }

    private static func host(of destination: String) -> String {
        String(String.UnicodeScalarView(destination.unicodeScalars.drop(while: { $0 != "@" }).dropFirst()))
    }

    /// The first problem with the fields, checked in `pairingProblem`'s order.
    var problem: String? {
        if !Self.isName(user, maxLength: 64) { return Problem.user }
        if !Self.isName(host, maxLength: 253) { return Problem.host }
        if port != 0 && !(1...65535).contains(port) { return Problem.port }
        if fingerprints.isEmpty { return Problem.noFingerprint }
        if !fingerprints.allSatisfy(Self.isFingerprint) { return Problem.fingerprint }
        if Set(fingerprints).count > Self.maxFingerprints { return Problem.tooManyFingerprints }
        return nil
    }

    private static let digits = Set("0123456789".unicodeScalars)
    private static let hex = Set("0123456789ABCDEFabcdef".unicodeScalars)
    private static let nameStart = Set("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_".unicodeScalars)
    private static let base64 = nameStart.subtracting(["_"]).union(["+", "/"])
    // The last of 43 base64 characters holds four bits of a 32-byte digest.
    private static let base64Last = Set("AEIMQUYcgkosw048".unicodeScalars)

    /// ASCII, not starting with "-" or ".", so it cannot become an ssh option.
    private static func isName(_ text: String, maxLength: Int) -> Bool {
        let scalars = Array(text.unicodeScalars)
        guard let first = scalars.first, nameStart.contains(first), scalars.count <= maxLength else { return false }
        return scalars.allSatisfy { nameStart.contains($0) || $0 == "." || $0 == "-" }
    }

    private static func isFingerprint(_ text: String) -> Bool {
        let scalars = Array(text.unicodeScalars)
        guard scalars.count == 50, string(scalars[..<7]) == "SHA256:", let last = scalars.last else { return false }
        return scalars[7..<49].allSatisfy { base64.contains($0) } && base64Last.contains(last)
    }

    private static func versionNumber(_ text: String) -> Int? {
        let scalars = Array(text.unicodeScalars)
        guard (1...6).contains(scalars.count), scalars.allSatisfy(digits.contains), scalars[0] != "0" else { return nil }
        return Int(text)
    }

    /// -1 for text that is not a port, so `problem` reports it in its turn.
    private static func portField(_ text: String?) -> Int {
        guard let text else { return 0 }
        let scalars = Array(text.unicodeScalars)
        guard (1...5).contains(scalars.count), scalars.allSatisfy(digits.contains), let port = Int(text),
            (1...65535).contains(port)
        else { return -1 }
        return port == 22 ? 0 : port
    }

    /// Only escapes of ASCII bytes, like the TypeScript reader: every field is
    /// ASCII, so a byte above 0x7f is damage.
    private static func percentDecoded(_ raw: String) -> String? {
        let scalars = Array(raw.unicodeScalars)
        var out = String.UnicodeScalarView()
        var at = 0
        while at < scalars.count {
            guard scalars[at] == "%" else {
                out.append(scalars[at])
                at += 1
                continue
            }
            // The digits are checked first because `UInt8(_:radix:)` also takes
            // a sign, and would read "%+F" as 15.
            guard at + 2 < scalars.count, hex.contains(scalars[at + 1]), hex.contains(scalars[at + 2]),
                let byte = UInt8(string(scalars[(at + 1)...(at + 2)]), radix: 16), byte <= 0x7f
            else { return nil }
            out.append(Unicode.Scalar(byte))
            at += 3
        }
        return String(out)
    }

    private static func trimmed(_ scalars: [Unicode.Scalar]) -> ArraySlice<Unicode.Scalar> {
        let blank: Set<Unicode.Scalar> = [" ", "\t", "\r", "\n"]
        guard let start = scalars.firstIndex(where: { !blank.contains($0) }),
            let end = scalars.lastIndex(where: { !blank.contains($0) })
        else { return [] }
        return scalars[start...end]
    }

    private static func asciiLowercased(_ text: String) -> String {
        var out = String.UnicodeScalarView()
        for scalar in text.unicodeScalars {
            out.append(("A"..."Z").contains(scalar) ? Unicode.Scalar(scalar.value + 32)! : scalar)
        }
        return String(out)
    }

    private static func string<S: Sequence>(_ scalars: S) -> String where S.Element == Unicode.Scalar {
        var out = String.UnicodeScalarView()
        out.append(contentsOf: scalars)
        return String(out)
    }
}

File: android/app/src/main/kotlin/sh/ledge/android/PairingCode.kt
package sh.ledge.android

/**
 * A pairing code: the server to dial and the host keys it offers, read from a
 * scanned or tapped link. The rules and their order are `shared/pairing.ts`'s,
 * as ios/Sources/PairingCode.swift takes them, and PairingCodeTest holds this
 * file to the same vectors (remote.md §4b). It imports nothing from Android, so
 * that test runs on the JVM.
 *
 * Every field a code carries is ASCII, so walking UTF-16 units here gives the
 * answers Swift's walk over Unicode scalars gives: a non-ASCII unit fails the
 * same check a non-ASCII scalar does.
 */
data class PairingCode(
    val user: String,
    /** A host name or an IPv4 address. */
    val host: String,
    /** 0 for sshd's default port, the way `ServerRecord.port` stores it. */
    val port: Int,
    /** `SHA256:…` as `ssh-keygen -lf` prints it, one per host key. */
    val fingerprints: List<String>,
) {
    val destination: String get() = "$user@$host"

    sealed interface Read {
        data class Code(val code: PairingCode) : Read
        data class Problem(val problem: String) : Read
    }

    /** A stored server, reduced to what the pin rule reads. */
    data class Known(
        val id: String,
        val destination: String,
        val port: Int,
        /** The pinned key's `SHA256:…`, or "" for a record whose pin was dropped. */
        val fingerprint: String,
    )

    /** What the stored servers make of this code (remote.md §4b, the first rule). */
    sealed interface Match {
        /** No record for this account at this host and port. The key the server
         * offers has to be one the code names, and it is pinned. */
        data object New : Match
        /** This account's record, pinned to a key the code names. The pin
         * decides the dial, and nothing is pinned again. */
        data class Pinned(val id: String) : Match
        /** This account's record with its pin dropped. The code's key is pinned. */
        data class Unpinned(val id: String) : Match
        /** A record at this host and port is pinned to a key the code does not
         * name. The code is refused, the way a changed host key is. */
        data class Conflict(val pinned: String) : Match
    }

    /**
     * A host key belongs to a host and a port rather than to an account, so a
     * pin on any account there can refuse the code. A host name compares
     * without regard to ASCII case, since DNS ignores it.
     */
    fun match(known: List<Known>): Match {
        val host = asciiLowercased(host)
        val here = known.filter { asciiLowercased(it.destination.substringAfter('@', "")) == host && it.port == port }
        here.firstOrNull { it.fingerprint.isNotEmpty() && it.fingerprint !in fingerprints }?.let { return Match.Conflict(it.fingerprint) }
        val mine = here.firstOrNull { it.destination.substringBefore('@') == user } ?: return Match.New
        return if (mine.fingerprint.isEmpty()) Match.Unpinned(mine.id) else Match.Pinned(mine.id)
    }

    /** The first problem with the fields, checked in `pairingProblem`'s order. */
    val problem: String?
        get() = when {
            !isName(user, 64) -> Problems.USER
            !isName(host, 253) -> Problems.HOST
            port != 0 && port !in 1..65535 -> Problems.PORT
            fingerprints.isEmpty() -> Problems.NO_FINGERPRINT
            !fingerprints.all(::isFingerprint) -> Problems.FINGERPRINT
            fingerprints.toSet().size > MAX_FINGERPRINTS -> Problems.TOO_MANY_FINGERPRINTS
            else -> null
        }

    object Problems {
        const val NOT_A_CODE = "This is not a Ledge pairing code."
        const val NEWER = "This pairing code needs a newer version of Ledge. Update the app, then try the code again."
        const val VERSION = "The pairing code's version is not one Ledge recognizes."
        const val ENCODING = "The pairing code has a damaged character in it."
        const val USER = "The pairing code has no valid account name."
        const val HOST = "The pairing code has no valid host name or IP address."
        const val PORT = "The pairing code's port is not a whole number from 1 to 65535."
        const val NO_FINGERPRINT = "The pairing code has no host key fingerprint."
        const val FINGERPRINT = "The pairing code has a host key fingerprint that is not a SHA256 fingerprint."
        const val TOO_MANY_FINGERPRINTS = "The pairing code names more than $MAX_FINGERPRINTS host keys."
        fun repeated(field: String) = "The pairing code gives the $field more than once."
    }

    companion object {
        const val VERSION = 1
        const val MAX_FINGERPRINTS = 4

        fun read(text: String): Read {
            val trimmed = text.trim(' ', '\t', '\r', '\n')
            val hash = trimmed.indexOf('#')
            if (hash < 0) return Read.Problem(Problems.NOT_A_CODE)
            val base = asciiLowercased(trimmed.substring(0, hash))
            if (base != "https://ledge.sh/pair" && base != "ledge://pair") return Read.Problem(Problems.NOT_A_CODE)

            val fields = trimmed.substring(hash + 1).split('&').filter { it.isNotEmpty() }.map { segment ->
                val eq = segment.indexOf('=')
                if (eq < 0) segment to "" else segment.substring(0, eq) to segment.substring(eq + 1)
            }
            fun raws(key: String) = fields.filter { it.first == key }.map { it.second }

            val versions = raws("v")
            if (versions.isEmpty()) return Read.Problem(Problems.NOT_A_CODE)
            if (versions.size > 1) return Read.Problem(Problems.repeated("version"))
            val version = versionNumber(versions[0]) ?: return Read.Problem(Problems.VERSION)
            if (version > VERSION) return Read.Problem(Problems.NEWER)

            for ((key, name) in listOf("u" to "account name", "h" to "host", "p" to "port")) {
                if (raws(key).size > 1) return Read.Problem(Problems.repeated(name))
            }
            val decoded = HashMap<String, List<String>>()
            for (key in listOf("u", "h", "p", "k")) {
                val values = raws(key).map(::percentDecoded)
                if (values.any { it == null }) return Read.Problem(Problems.ENCODING)
                decoded[key] = values.filterNotNull()
            }

            val code = PairingCode(
                user = decoded["u"]!!.firstOrNull().orEmpty(),
                host = decoded["h"]!!.firstOrNull().orEmpty(),
                port = portField(decoded["p"]!!.firstOrNull()),
                fingerprints = decoded["k"]!!.distinct(),
            )
            return code.problem?.let { Read.Problem(it) } ?: Read.Code(code)
        }

        private const val NAME_START = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_"
        private const val BASE64 = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
        // The last of 43 base64 characters holds four bits of a 32-byte digest.
        private const val BASE64_LAST = "AEIMQUYcgkosw048"
        private const val HEX = "0123456789ABCDEFabcdef"

        /** ASCII, not starting with "-" or ".", so it cannot become an ssh option. */
        private fun isName(text: String, maxLength: Int): Boolean =
            text.isNotEmpty() && text.length <= maxLength && text[0] in NAME_START &&
                text.all { it in NAME_START || it == '.' || it == '-' }

        private fun isFingerprint(text: String): Boolean =
            text.length == 50 && text.startsWith("SHA256:") &&
                text.substring(7, 49).all { it in BASE64 } && text[49] in BASE64_LAST

        private fun versionNumber(text: String): Int? =
            if (text.length in 1..6 && text.all { it in '0'..'9' } && text[0] != '0') text.toInt() else null

        /** -1 for text that is not a port, so `problem` reports it in its turn. */
        private fun portField(text: String?): Int {
            if (text == null) return 0
            if (text.length !in 1..5 || !text.all { it in '0'..'9' }) return -1
            val port = text.toInt()
            if (port !in 1..65535) return -1
            return if (port == 22) 0 else port
        }

        /** Only escapes of ASCII bytes, like the TypeScript reader: every field
         * is ASCII, so a byte above 0x7f is damage. */
        private fun percentDecoded(raw: String): String? {
            val out = StringBuilder()
            var at = 0
            while (at < raw.length) {
                if (raw[at] != '%') {
                    out.append(raw[at])
                    at += 1
                    continue
                }
                if (at + 2 >= raw.length || raw[at + 1] !in HEX || raw[at + 2] !in HEX) return null
                val byte = raw.substring(at + 1, at + 3).toInt(16)
                if (byte > 0x7f) return null
                out.append(byte.toChar())
                at += 3
            }
            return out.toString()
        }

        private fun asciiLowercased(text: String): String =
            buildString { text.forEach { append(if (it in 'A'..'Z') it + 32 else it) } }
    }
}
